pkg/network/ovn_kubernetes.go

```diff
@@ -833,6 +849,22 @@ func bootstrapOVNConfig(conf *operv1.Network, kubeClient cnoclient.Client, hc *h
 		return nil, fmt.Errorf("Node %s has multiple hardware offload labels.", nodeName)
 	}
 
+	dmc = types.NamespacedName{Namespace: names.APPLIED_NAMESPACE, Name: names.APPLIED_PREFIX + names.OPERATOR_CONFIG}
+	err = kubeClient.ClientFor("").CRClient().Get(context.TODO(), dmc, cm)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("Could not determine stale masquerade subnets: %w", err)
+		}
+	} else {
+		spec := &operv1.NetworkSpec{}
+		err = json.Unmarshal([]byte(cm.Data["applied"]), spec)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to unmarshal data from configmap %s in %s namespace: %w", names.APPLIED_PREFIX+names.OPERATOR_CONFIG, names.APPLIED_NAMESPACE, err)
+		}
+		ovnConfigResult.V4OldMasqueradeSubnet = spec.DefaultNetwork.OVNKubernetesConfig.GatewayConfig.IPv4.InternalMasqueradeSubnet
```

Comment:
Have you tested that this change works? I am concerned that CNO will cleanup the `V[4/6]OldMasqueradeSubnet` on the next reconciliation and that might happen before ovn-k has a chance to cleanup the stale entries.

Reply:
yaa, as @trozet mentioned at https://github.com/ovn-org/ovn-kubernetes/pull/4496#issuecomment-2251451338 , we need to document that user should refrain from changing masquerade subnet again until all ovnkube-node PODs have rolled out upon changing once. User should also refrain from changing masquerade subnet before completion of cluster upgrade.

If user is not changing masquerade subnet before rollout of all ovnkube-node PODs then there should not be a problem as first bootstrap happens then we run Render and then we apply new config to the applied-cluster config map.
https://github.com/openshift/cluster-network-operator/blob/master/pkg/controller/operconfig/operconfig_controller.go#L368-L417

Reply:
Hmm, I understand your concern. very nice catch :+1: . I think we need to annotate the daemonset with stale masquerade subnet only with valid IP in that case. wdyt?

Reply:
@kyrtapz I am planning following to tackle this problem:

- Instead of updating 008-script-lib.yaml with new and stale masquerade subnet, I am going to update 004-config.yaml. In that case ovnkube-node daemonset is not going to be updated directly when masquerade subnet is changed at day 2. Instead ovnkube-config configmap will be updated.
- Since changes to ovnkube-config CM does not rollout ovnkube daemonset, I am going to add an annotation to the daemonset with a bool value to rollout ovnkube-node only when V[4][6]StaleMasqueradeSubnet is not empty.

If consecutive reconciliation happens then it is not going to update the daemonset and would stop a new rollout. In documentation, we would also need to mention that v4/v6 masquerade subnet should not be changed again until last rollout of ovnkube-node completes. 

Do you have any other suggestions? Thanks in advance!

Reply:
Closing this conversation as we are not anymore passing on old masquerade subnet to ovnkube.